numpy/core/src/umath/ufunc_object.c

```diff
@@ -5347,11 +5489,59 @@ PyUFunc_RegisterLoopForType(PyUFuncObject *ufunc,
     if (arg_types != NULL) {
         for (i = 0; i < ufunc->nargs; i++) {
             newtypes[i] = arg_types[i];
+            signature[i] = PyArray_DTypeFromTypeNum(arg_types[i]);
+            Py_DECREF(signature[i]);  /* DType can't be deleted... */
         }
     }
     else {
         for (i = 0; i < ufunc->nargs; i++) {
             newtypes[i] = usertype;
+            signature[i] = PyArray_DTypeFromTypeNum(usertype);
+            Py_DECREF(signature[i]);  /* DType can't be deleted... */
+        }
+    }
+
+    signature_tuple = PyArray_TupleFromItems(
+            ufunc->nargs, (PyObject **)signature, 0);
+    if (signature_tuple == NULL) {
+        goto fail;
+    }
+    /*
+     * We add the loop to the new list of all loops and promoters.  If the
+     * equivalent loop was already added, skip this.  If a different loop is
```

Comment:
I'm a bit confused why we skip if the equivalent loop was already added. Shouldn't we replace? Or check that the implementation is identical? Or is that to be done later so that it replaces the still existing registration?

Reply:
Maybe this comment now makes it clear?:

>We add the loop to the list of all loops and promoters.  If the equivalent loop was already added, skip this.
> Note that even then the ufunc is still modified: The legacy ArrayMethod already looks up the inner-loop from the ufunc (and this is replaced below!).
> If the existing one is not a legacy ArrayMethod, we raise currently: A new-style loop should not be replaced by an old-style one.

although, we may have to allow that replacement with a deprecation-warning before too long, so that we don't break those replacing our loops but can use them internally anyway.